### PR TITLE
CAMEL-20565 - Azure servicebus dead-lettering

### PR DESCRIPTION
**Pull request by romain-pfund**
# Description

Add a enableDeadLettering consumer configuration to allow deadlettering failed message to the deadletter sub-queue

According to [pplication-level-dead-lettering](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-dead-letter-queues#application-level-dead-lettering). The type of the exception is in the DeadLetterReason and the stack trace of the exception in the DeadLetterDescription.

# Target

- [x] I checked that the commit is targeting the correct branch (note that Camel 3 uses `camel-3.x`, whereas Camel 4 uses the `main` branch)

# Tracking
- [x] If this is a large change, bug fix, or code improvement, I checked there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).

# Apache Camel coding standards and style

- [ ] I checked that each commit in the pull request has a meaningful subject line and body.

- [ ] I have run `mvn clean install -DskipTests` locally and I have committed all auto-generated changes

### CODE REVIEW

File: components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java [hunk 1]
```diff
@@ -228,8 +232,21 @@ public void onFailure(Exchange exchange) {
             if (cause != null) {
                 getExceptionHandler().handleException("Error during processing exchange.", exchange, cause);
             }
+
             if (!getConfiguration().isDisableAutoComplete()) {
-                clientWrapper.abandon(message).subscribeOn(Schedulers.boundedElastic()).subscribe();
+                if (getConfiguration().isEnableDeadLettering()) {
```
Comment (oscerd) [inline components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java:237]: I think you need also to double check if there is actually a subqueue in your client instance.

**Review by oscerd — CHANGES_REQUESTED** [review]
Overall looks good, we need to do a full build and I think there is one thing missing about checking the subqueue

File: components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java (hunk 1 above)
Comment (romain-pfund) [reply components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java:237]: Yes your're right, thanks for the feedback.

Comment (romain-pfund) [reply components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java:237]: I've just removed my precedent comment as it was not correct.

According to the [Azure ServiceBus documentation](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-dead-letter-queues)
"Azure Service Bus queues and topic subscriptions provide a secondary subqueue, called a dead-letter queue"
So every queue or subscription has a deadletter.

but you are right, is the consumer is already consuming subqueue, we cannot deadletter the message. So I need to check if the subQueue option is set

Comment (oscerd) [reply components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java:237]: Yes, that was what I meant.

Comment (romain-pfund) [reply components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java]: I've added the check
`Objects.isNull(getConfiguration().getSubQueue()) || getConfiguration().getSubQueue().equals(SubQueue.NONE)`

**Comment by romain-pfund** [comment]
Is there a way to write a test to cover this "onFailure" part ?

for now I'm manully testing with route like:
```
  public static final String AZURE_SERVICE_BUS_COMPONENT = "azure-servicebus";
  @Override
  public void configure() {
    String subscription = "test-scr";
    String inputTopic = "test";
    fromF("%s:%s?subscriptionName=%s&enableDeadLettering=true", AZURE_SERVICE_BUS_COMPONENT, inputTopic, subscription)
        .routeId("routeTestId")
        .throwException(new Exception("test failure"))
        .log("${body}");

  }
```

File: components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java [hunk 2]
```diff
@@ -228,8 +233,23 @@ public void onFailure(Exchange exchange) {
             if (cause != null) {
                 getExceptionHandler().handleException("Error during processing exchange.", exchange, cause);
             }
+
             if (!getConfiguration().isDisableAutoComplete()) {
-                clientWrapper.abandon(message).subscribeOn(Schedulers.boundedElastic()).subscribe();
+                if (getConfiguration().isEnableDeadLettering() && (Objects.isNull(getConfiguration().getSubQueue()) ||
```
Comment (oscerd) [inline components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java:238]: You can use also ObjectHelper from camel-utils to be aligned with the codebase

Comment (romain-pfund) [reply components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java:238]: ok thanks I'll check.
First time here so happy discover new things

**Comment by oscerd** [comment]
> Is there a way to write a test to cover this "onFailure" part ?
> 
> for now I'm manully testing with route like:
> 
> ```
>   public static final String AZURE_SERVICE_BUS_COMPONENT = "azure-servicebus";
>   @Override
>   public void configure() {
>     String subscription = "test-scr";
>     String inputTopic = "test";
>     fromF("%s:%s?subscriptionName=%s&enableDeadLettering=true", AZURE_SERVICE_BUS_COMPONENT, inputTopic, subscription)
>         .routeId("routeTestId")
>         .throwException(new Exception("test failure"))
>         .log("${body}");
> 
>   }
> ```

In this component we are only using integration tests, so you could have a look at ServiceBusConsumerTest and write a similar tests with this option enabled, but you'll need a real Azure Servicebus queue/topic.

File: components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java (hunk 2 above)
Comment (romain-pfund) [reply components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java:238]: so something like
`(ObjectHelper.isEmpty(getConfiguration().getSubQueue()) ||ObjectHelper.equal(getConfiguration().getSubQueue(), SubQueue.NONE)` ?

Comment (oscerd) [reply components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java:238]: Yes, something like that.

File: components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java (hunk 1 above)
Comment (romain-pfund) [reply components/camel-azure/camel-azure-servicebus/src/main/java/org/apache/camel/component/azure/servicebus/ServiceBusConsumer.java]: it's pushed

**Comment by romain-pfund** [comment]
just a question, it seams that the PR build is looping, is it normal

**Comment by oscerd** [comment]
It is not looping, it is just that you need to run a full build to regenerate the metadata of Azure Service bus component. So you should run mvn clean install -DskipTests from the root folder of the project and update the PR with a new commit.

**Review by oscerd — APPROVED** [review]
This looks good by the way. We can merge and regenerate after merging

**Comment by romain-pfund** [comment]
I want just do the integration test. I should have time to do it tomorrow or monday

**Comment by oscerd** [comment]
No rush. Thanks for your time!

**Comment by romain-pfund** [comment]
Didn't had time yesterday, Maybe we can look to merge it and I'll do the intégration test later in another PR.
I'm a bit full this week, may have more time next week

**Comment by oscerd** [comment]
Sure, please make the PR ready to review.

**Comment by romain-pfund** [comment]
shoudn't I do the mvn clean install -DskipTests and commit first ?

**Comment by oscerd** [comment]
Yes, if you have time.

**Comment by romain-pfund** [comment]
got trouble building it (i'm on a windows os right now), you said we can merge then rebuild ?

**Comment by oscerd** [comment]
Let's merge this and regen

**Pull request merged**